### PR TITLE
Avoid array temporaries

### PR DESCRIPTION
**Pull request by matt-frey**
Fortran creates temporary arrays when passing sub-ranges to subroutines.

### CODE REVIEW

File: src/3d/inversion/inversion.f90 [hunk 1]
```diff
@@ -360,10 +360,11 @@ end subroutine vorticity_tendency
 
         !::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
 
-        ! Note: f is overwritten
+        ! Note: f is overwritten; only the range div(0:nz, :, :) contains
+        ! valid data
         subroutine divergence(f, div)
             double precision, intent(inout) :: f(-1:nz+1, box%hlo(2):box%hhi(2), box%hlo(1):box%hhi(1), n_dim)
-            double precision, intent(out)   :: div(0:nz, box%hlo(2):box%hhi(2), box%hlo(1):box%hhi(1))
+            double precision, intent(out)   :: div(-1:nz+1, box%hlo(2):box%hhi(2), box%hlo(1):box%hhi(1))
```
Comment (sjboeing) [inline src/3d/inversion/inversion.f90:367]: Should there be default output for div at -1 or nz+1 (if only for safety)? I am not sure about this.

File: src/3d/inversion/inversion_utils.f90 [hunk 2]
```diff
@@ -482,6 +483,32 @@ subroutine central_diffz(fs, ds)
...
 
+        end subroutine central_diffz_semi_spectral
+
+        !::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+
+        !Calculates df/dz for a field f using 2nd-order differencing.
+        !Here df = df/dz. In physical space.
+        subroutine central_diffz(f, df)
+            double precision, intent(in)  :: f(-1:nz+1,  box%hlo(2):box%hhi(2), box%hlo(1):box%hhi(1))
+            double precision, intent(out) :: df(-1:nz+1, box%hlo(2):box%hhi(2), box%hlo(1):box%hhi(1))
```
Comment (sjboeing) [inline src/3d/inversion/inversion_utils.f90:494]: Sam question here.

**Review by sjboeing — COMMENTED** [review]
Thanks for writing this Matthias! I can see why you initially went for the other approach, but thought it would be worth trying to get rid of the compiler warnings. Not quite sure whether to do anything (by default) with the values of the output fields at z=-1 and z=nz+1.

File: src/3d/inversion/inversion.f90 (hunk 1 above)
Comment (matt-frey) [reply src/3d/inversion/inversion.f90:367]: We can use linear extrapolation.

**Comment by matt-frey** [comment]
We could also get rid of these warnings by explicitly creating a temporary variable ourselves.

**Comment by sjboeing** [comment]
What do you think is best practice/most efficient? I would be OK with either solution. Guess this is the next one to merge after the period BCs PR.

**Comment by matt-frey** [comment]
Avoiding temporary arrays is definitely more efficient as we do not copy memory. The indices `-1` and `nz+1` and we can either set them to zero or (maybe better) use linear extrapolation.

**Comment by matt-frey** [comment]
My suggested change: https://github.com/EPIC-model/epic/pull/504/commits/5759bceae46483e0074891346e54e4fa4acc5bb6

**Comment by sjboeing** [comment]
Let's use the extrapolation, which I agree should be safer.

**Pull request merged**